In [243]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [244]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
import statsmodels.formula.api as smf

## Обработка датасета

In [245]:
data = pd.read_csv('/content/drive/MyDrive/CarPrice_Assignment.csv')
data.head()

car_ID  symboling                   CarName  ... citympg highwaympg    price
0       1          3        alfa-romero giulia  ...      21         27  13495.0
1       2          3       alfa-romero stelvio  ...      21         27  16500.0
2       3          1  alfa-romero Quadrifoglio  ...      19         26  16500.0
3       4          2               audi 100 ls  ...      24         30  13950.0
4       5          2                audi 100ls  ...      18         22  17450.0

[5 rows x 26 columns]

In [246]:
data.set_index('car_ID', inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 205 entries, 1 to 205
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   symboling         205 non-null    int64  
 1   CarName           205 non-null    object 
 2   fueltype          205 non-null    object 
 3   aspiration        205 non-null    object 
 4   doornumber        205 non-null    object 
 5   carbody           205 non-null    object 
 6   drivewheel        205 non-null    object 
 7   enginelocation    205 non-null    object 
 8   wheelbase         205 non-null    float64
 9   carlength         205 non-null    float64
 10  carwidth          205 non-null    float64
 11  carheight         205 non-null    float64
 12  curbweight        205 non-null    int64  
 13  enginetype        205 non-null    object 
 14  cylindernumber    205 non-null    object 
 15  enginesize        205 non-null    int64  
 16  fuelsystem        205 non-null    object 
 1

In [247]:
data.symboling.value_counts()

 0    67
 1    54
 2    32
 3    27
-1    22
-2     3
Name: symboling, dtype: int64

In [248]:
data.CarName.value_counts()
# Это не Company name, придётся выбирать навзания компаний вручныю

toyota corolla               6
peugeot 504                  6
toyota corona                6
subaru dl                    4
mazda 626                    3
                            ..
jaguar xf                    1
dodge coronet custom (sw)    1
dodge monaco (sw)            1
subaru baja                  1
toyota corolla tercel        1
Name: CarName, Length: 147, dtype: int64

In [249]:
# Надо найти список всех брендов
brand_car_list = []

import requests
res = requests.get('https://private-anon-03814797ef-carsapi1.apiary-mock.com/manufacturers')
for el in res.json():
    brand_car_list.append(el['name'])


In [250]:
for brand in brand_car_list:
    data['CarName'].loc[data[data.CarName.apply(lambda x: brand.lower() in x)].index] = brand.lower()


/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [251]:
data.CarName.value_counts()

toyota                            31
nissan                            17
mazda                             15
mitsubishi                        13
honda                             13
subaru                            12
volvo                             11
volkswagen                         9
buick                              8
dodge                              8
bmw                                8
audi                               7
peugeot 504                        6
porsche                            4
chevrolet                          3
jaguar                             3
saab 99gle                         2
plymouth fury iii                  2
isuzu D-Max                        2
saab 99le                          2
peugeot 604sl                      2
saab 99e                           2
vokswagen rabbit                   1
alfa-romero stelvio                1
porcshce panamera                  1
isuzu MU-X                         1
mercury cougar                     1
p

### Очень длинная колонка со списком брендов (скорпировано с гита)

In [252]:
brand_car_list_2 = [
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/abarth1.png",
    "name": "Abarth"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ac-cars.png",
    "name": "AC"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/acura.png",
    "name": "Acura"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/aixam.png",
    "name": "Aixam"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/alfa_romeo.png",
    "name": "Alfa Romeo"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ariel.png",
    "name": "Ariel"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/10/arrinera-logo.png",
    "name": "Arrinera"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/aston_martin.png",
    "name": "Aston Martin"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/audi.png",
    "name": "Audi"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/bentley.png",
    "name": "Bentley"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/bmw.png",
    "name": "BMW"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/bugatti.png",
    "name": "Bugatti"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/buick.png",
    "name": "Buick"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/cadillac.png",
    "name": "Cadillac"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/caterham.png",
    "name": "Caterham"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/chevrolet.png",
    "name": "Chevrolet"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/chrysler.png",
    "name": "Chrysler"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/citroen.png",
    "name": "Citroën"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/corvette.png",
    "name": "Corvette"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/dacia.png",
    "name": "Dacia"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/daf.png",
    "name": "Daf"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/daihat.png",
    "name": "Daihatsu"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/dodge.png",
    "name": "Dodge"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/10/drmotor.png",
    "name": "DR Motor"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/elfin.png",
    "name": "Elfin"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ferrari.png",
    "name": "Ferrari"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/fiat.png",
    "name": "Fiat"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ford.png",
    "name": "Ford"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/gaz.png",
    "name": "Gaz"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/geely.png",
    "name": "Geely"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/gillet.png",
    "name": "Gillet"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ginetta.png",
    "name": "Ginetta"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/gm.png",
    "name": "General Motors"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/gmc.png",
    "name": "GMC"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2013/10/Great-Wall.png",
    "name": "Great Wall"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/gumpert.png",
    "name": "Gumpert"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/10/hennessey.png",
    "name": "Hennessey logo"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/10/holden.png",
    "name": "Holden"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/honda.png",
    "name": "Honda"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/hummer.png",
    "name": "Hummer"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/hyundai.png",
    "name": "Hyundai"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/inf.png",
    "name": "Infiniti"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/isuzu.png",
    "name": "Isuzu"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/jagu.png",
    "name": "Jaguar"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/jeep.png",
    "name": "Jeep"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/joss.png",
    "name": "Joss"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/kia.png",
    "name": "Kia"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/koenigsegg.png",
    "name": "Koenigsegg"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/lada.png",
    "name": "Lada"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/lamb.png",
    "name": "Lamborghini"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/lancia.png",
    "name": "Lancia"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/land-rover.png",
    "name": "Land Rover"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/lexus.png",
    "name": "Lexus"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/linc.png",
    "name": "Lincoln"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/lotus.png",
    "name": "Lotus"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/11/luxgen-logo.png",
    "name": "Luxgen"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/11/mahindra.png",
    "name": "Mahindra"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/12/Maruti_Suzuki.png",
    "name": "Maruti Suzuki"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mase.png",
    "name": "Maserati"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/maybach.png",
    "name": "Maybach"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mazda.png",
    "name": "Mazda"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mclaren.png",
    "name": "Mclaren"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/marchedrs.png",
    "name": "Mercedes"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mg.png",
    "name": "MG"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mini.png",
    "name": "Mini"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mitub.png",
    "name": "Mitsubishi"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/morgan.png",
    "name": "Morgan Motor"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/mustang.png",
    "name": "Mustang logo"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/nissan.png",
    "name": "Nissan"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/noble.png",
    "name": "Noble"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/opel.png",
    "name": "Opel"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/pagani.png",
    "name": "Pagani"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/panoz.png",
    "name": "Panoz"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/perodua.png",
    "name": "Perodua"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/peug.png",
    "name": "Peugeot"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/piaggio.png",
    "name": "Piaggio"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/pininfarina.png",
    "name": "Pininfarina"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/porsche.png",
    "name": "Porsche"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/proton.png",
    "name": "Proton"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/renault.png",
    "name": "Renault"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/reva.png",
    "name": "Reva"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2012/01/rimac-150x150.png",
    "name": "Rimac"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/rolls-royce.png",
    "name": "Rolls Royce"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ruf.png",
    "name": "Ruf logo"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/saab.png",
    "name": "Saab"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/scania.png",
    "name": "Scania"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/scion.png",
    "name": "Scion"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/seat.png",
    "name": "Seat"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/10/shelby.png",
    "name": "Shelby"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/skoda.png",
    "name": "Skoda"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/smart.png",
    "name": "Smart"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/spyker.png",
    "name": "Spyker"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/ssangyong.png",
    "name": "Ssangyong"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/10/ssc.png",
    "name": "SSC"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/subaru.png",
    "name": "Subaru"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/suzuki.png",
    "name": "Suzuki"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/tata.png",
    "name": "Tata"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/tatra.png",
    "name": "Tatra"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/tesla.png",
    "name": "Tesla"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/toyota.png",
    "name": "Toyota"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/1176359_187686584745478_1792749640_n.jpg",
    "name": "Tramontana"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/troller.png",
    "name": "Troller"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/tvr.png",
    "name": "TVR"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/uaz.png",
    "name": "UAZ"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/vandenbrink_design.png",
    "name": "Vandenbrink"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/vauxhall.png",
    "name": "Vauxhall"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/vector_motors.png",
    "name": "Vector Motors"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/venturi.png",
    "name": "Venturi"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/volkswagen.png",
    "name": "Volkswagen"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/volvo.png",
    "name": "Volvo"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/wiesmann.png",
    "name": "Wiesmann"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/zagato.png",
    "name": "Zagato"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/zaz.png",
    "name": "Zaz"
  },
  {
    "logo": "https://www.car-logos.org/wp-content/uploads/2011/09/zil.png",
    "name": "Zil"
  }
]

### Конец длинной колонки

In [253]:
for brand in brand_car_list_2:
    data['CarName'].loc[data[data.CarName.apply(lambda x: brand['name'].lower() in x)].index] = brand['name'].lower()


/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [254]:
data.CarName.value_counts()[data.CarName.value_counts() < 3]

renault                           2
plymouth fury iii                 2
maxda glc deluxe                  1
alfa-romero stelvio               1
vokswagen rabbit                  1
porcshce panamera                 1
mercury cougar                    1
alfa-romero Quadrifoglio          1
ram                               1
maxda rx3                         1
Nissan versa                      1
plymouth satellite custom (sw)    1
plymouth cricket                  1
alfa-romero giulia                1
plymouth duster                   1
vw dasher                         1
plymouth fury gran sedan          1
vw rabbit                         1
toyouta tercel                    1
plymouth valiant                  1
Name: CarName, dtype: int64

Остатки можно вручную доделать

In [255]:
data['CarName'].loc[data[data.CarName.apply(lambda x: 'plymouth' in x)].index] = 'plymouth'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'maxda' in x)].index] = 'mazda'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'vw' in x)].index] = 'volkswagen'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'vokswagen' in x)].index] = 'volkswagen'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'alfa-romero' in x)].index] = 'alfa-romero'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'toyouta' in x)].index] = 'toyota'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'porcshce' in x)].index] = 'porsche'
data['CarName'].loc[data[data.CarName.apply(lambda x: 'Nissan' in x)].index] = 'nissan'


/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [256]:
data.CarName.value_counts()
# Важный показатель для оценки стоимости, но очень несбалансированны классы. Возможно будет лучше убрать эту колонку

toyota            32
nissan            18
mazda             17
honda             13
mitsubishi        13
subaru            12
volkswagen        12
volvo             11
peugeot           11
bmw                8
dodge              8
buick              8
audi               7
plymouth           7
saab               6
porsche            5
isuzu              4
alfa-romero        3
chevrolet          3
jaguar             3
renault            2
ram                1
mercury cougar     1
Name: CarName, dtype: int64

In [257]:
for cat in data.columns:
    print(cat, len(data[cat].value_counts()))
# Посмотрим над какимим категориальными фичами стоит поработать (над теми, у которых слишком много классов)

symboling 6
CarName 23
fueltype 2
aspiration 2
doornumber 2
carbody 5
drivewheel 3
enginelocation 2
wheelbase 53
carlength 75
carwidth 44
carheight 49
curbweight 171
enginetype 7
cylindernumber 7
enginesize 44
fuelsystem 8
boreratio 38
stroke 37
compressionratio 32
horsepower 59
peakrpm 23
citympg 29
highwaympg 30
price 189


Остальные категориальные колонки в норме

Посмотрим на них

Необходимы три типа колонок: категориальные, порядковые (по типу количества цилиндров, т.к. они скорее всего коррелируют с ценой сами, без кодирования) и числовые

In [258]:
cat_cols = {'CarName', 'fueltype', 'aspiration', 'carbody', 'enginetype', 'fuelsystem', 'drivewheel', 'enginelocation'}
rang_cols = {'symboling', 'doornumber', 'cylindernumber'}
num_cols = set(data.columns) - cat_cols - rang_cols
num_cols.remove('price')

In [259]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import Normalizer


## Baseline

Только числовые колонки

In [260]:
X = data[num_cols]
y = data['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, shuffle=True, random_state=42)

transformer = Normalizer().fit(X_train)

reg = LinearRegression()

scores = cross_val_score(reg, transformer.transform(X_train), y_train, cv=5, scoring = 'neg_mean_absolute_error')
print((-1) * scores.mean())
preds = reg.fit(transformer.transform(X_train), y_train).predict(transformer.transform(X_test))
print(mean_absolute_error(preds, y_test))

2535.2064027997612
2717.2338087162293


Зафиксировали значения метрик.

    Для тестовой выборки mae = 2535
    Для обучающей = 2717

(тестовую   будем и обучающую менять не будем, представим, что ответов на тестовую у нас нет)

## Обработка числовых колонок

Воспользуемся методом наименьших квадратов и оценим значимость числовых признаков

In [261]:
resultsm = smf.ols(f'price ~ {"+".join(num_cols)}', data=data).fit()
resultsm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.851
Model:                            OLS   Adj. R-squared:                  0.841
Method:                 Least Squares   F-statistic:                     83.78
Date:                Sun, 21 Nov 2021   Prob (F-statistic):           1.68e-71
Time:                        22:14:46   Log-Likelihood:                -1937.5
No. Observations:                 205   AIC:                             3903.
Df Residuals:                     191   BIC:                             3949.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         -4.75e+04   1.53e+04     -3.111      0.002   -7.76e+04   -1.74e+04
stroke           -3034.6060    778.604     -3.897      0.000   -4570.373   -1498.839
carwidth           505.5716    246.013      2.055      0.041      20.319     990.824
boreratio        -1002.5654   1195.798     -0.838      0.403   -3361.231    1356.100
peakrpm              2.3751      0.671      3.540      0.001       1.052       3.698
carheight          163.1801    135.721      1.202      0.231    -104.524     430.884
curbweight           1.8846      1.737      1.085      0.279      -1.542       5.312
compressionratio   298.1369     82.914      3.596      0.000     134.592     461.682
citympg           -320.3545    177.769     -1.802      0.073    -670.998      30.289
carlength          -94.6752     55.557     -1.704      0.090    -204.259      14.909
enginesize         117.3461     13.837      8.481      0.000      90.054     144.638
wheelbase          122.6169    100.465      1.220      0.224     -75.546     320.780
horsepower          30.8086     16.216      1.900      0.059      -1.177      62.795
highwaympg         202.8221    159.760      1.270      0.206    -112.299     517.943
==============================================================================
Omnibus:                       24.541   Durbin-Watson:                   0.930
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               81.326
Skew:                           0.383   Prob(JB):                     2.19e-18
Kurtosis:                       5.989   Cond. No.                     3.94e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.94e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Будем убирать незначимые параметры по одному

In [262]:
useful_cols = num_cols - set(['boreratio', 'carheight', 'carheight', 'carlength', 'citympg', 'wheelbase', 'highwaympg'])

In [263]:
resultsm = smf.ols(f'price ~ {"+".join(useful_cols)}', data=data).fit()
resultsm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.845
Model:                            OLS   Adj. R-squared:                  0.840
Method:                 Least Squares   F-statistic:                     153.8
Date:                Sun, 21 Nov 2021   Prob (F-statistic):           2.88e-76
Time:                        22:14:46   Log-Likelihood:                -1941.2
No. Observations:                 205   AIC:                             3898.
Df Residuals:                     197   BIC:                             3925.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept        -4.988e+04   1.24e+04     -4.026      0.000   -7.43e+04   -2.54e+04
compressionratio   245.4669     68.175      3.601      0.000     111.020     379.913
stroke           -2882.0051    751.278     -3.836      0.000   -4363.585   -1400.425
enginesize         110.4078     13.047      8.462      0.000      84.679     136.137
peakrpm              2.5344      0.640      3.958      0.000       1.272       3.797
carwidth           513.9413    211.434      2.431      0.016      96.977     930.905
horsepower          31.2353     13.123      2.380      0.018       5.356      57.114
curbweight           2.3162      1.180      1.962      0.051      -0.012       4.644
==============================================================================
Omnibus:                       21.221   Durbin-Watson:                   0.845
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               62.429
Skew:                           0.344   Prob(JB):                     2.78e-14
Kurtosis:                       5.615   Cond. No.                     3.19e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.19e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [264]:
X = data[useful_cols]
y = data['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, shuffle=True, random_state=42)

transformer = Normalizer().fit(X_train)

reg = LinearRegression()

scores = cross_val_score(reg, transformer.transform(X_train), y_train, cv=5, scoring = 'neg_mean_absolute_error')
print((-1) * scores.mean())
preds = reg.fit(transformer.transform(X_train), y_train).predict(transformer.transform(X_test))
print(mean_absolute_error(preds, y_test))

2513.1570488642046
2566.0812019845707


Значения ошибки и на тестовой, и на обучающей снизились, т.к. мы избавились от незначимых признаков

## Добавление и обработка ранговых колонок

Теперь предобработаем ранговые колонки и добавим их

In [265]:
data[rang_cols].head()

doornumber  symboling cylindernumber
car_ID                                     
1             two          3           four
2             two          3           four
3             two          1            six
4            four          2           four
5            four          2           five

In [266]:
data.doornumber.loc[data[data.doornumber == 'two'].index] = 2
data.doornumber.loc[data[data.doornumber == 'four'].index] = 4
data.doornumber = data.doornumber.astype(int)

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [267]:
!pip install word2number
from word2number import w2n

In [268]:
data.cylindernumber = data.cylindernumber.apply(w2n.word_to_num).astype(int)

In [269]:
useful_cols_2 = useful_cols | rang_cols

In [270]:
resultsm = smf.ols(f'price ~ {"+".join(useful_cols_2)}', data=data).fit()
resultsm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.848
Model:                            OLS   Adj. R-squared:                  0.840
Method:                 Least Squares   F-statistic:                     107.9
Date:                Sun, 21 Nov 2021   Prob (F-statistic):           1.23e-73
Time:                        22:14:49   Log-Likelihood:                -1939.7
No. Observations:                 205   AIC:                             3901.
Df Residuals:                     194   BIC:                             3938.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept        -4.964e+04   1.25e+04     -3.978      0.000   -7.43e+04    -2.5e+04
compressionratio   261.3546     69.275      3.773      0.000     124.726     397.983
stroke           -3359.5035    809.120     -4.152      0.000   -4955.305   -1763.702
cylindernumber    -728.5383    472.807     -1.541      0.125   -1661.039     203.963
enginesize         134.3006     20.112      6.678      0.000      94.635     173.966
carwidth           547.7386    213.019      2.571      0.011     127.608     967.869
peakrpm              2.6684      0.649      4.114      0.000       1.389       3.948
horsepower          31.1819     13.802      2.259      0.025       3.960      58.404
doornumber          69.3270    315.643      0.220      0.826    -553.205     691.859
curbweight           1.5805      1.343      1.177      0.241      -1.069       4.229
symboling           89.1103    250.947      0.355      0.723    -405.825     584.046
==============================================================================
Omnibus:                       19.585   Durbin-Watson:                   0.812
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               45.208
Skew:                           0.403   Prob(JB):                     1.52e-10
Kurtosis:                       5.154   Cond. No.                     3.21e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.21e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Все три ранговых переменных оказываются незначимыми и ухудшают итоговое качество

In [271]:
X = data[useful_cols]
y = data['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, shuffle=True, random_state=42)

transformer = Normalizer().fit(X_train)

reg = LinearRegression()

scores = cross_val_score(reg, transformer.transform(X_train), y_train, cv=5, scoring = 'neg_mean_absolute_error')
print((-1) * scores.mean())
preds = reg.fit(transformer.transform(X_train), y_train).predict(transformer.transform(X_test))
print(mean_absolute_error(preds, y_test))

2513.1570488642046
2566.0812019845707


## Категориальные признаки

In [272]:
data_1 = data[useful_cols].join(pd.get_dummies(data[cat_cols]))

In [273]:
X = data_1
y = data['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, shuffle=True, random_state=42)

transformer = Normalizer().fit(X_train)

reg = LinearRegression()

scores = cross_val_score(reg, transformer.transform(X_train), y_train, cv=5, scoring = 'neg_mean_absolute_error')
print((-1) * scores.mean())
preds = reg.fit(transformer.transform(X_train), y_train).predict(transformer.transform(X_test))
print(mean_absolute_error(preds, y_test))

1839.7045099813015
2530.1834913367825


Качество на обучающей выросло заметно, но и на тестовой не упало